### PR TITLE
Implement update-version-in-file command

### PR DESCRIPTION
**Pull request by tedim52**
implements a command to execute logic in [`kurtosistech/developer-tools:update-version-in-file.sh`](https://github.com/kurtosis-tech/developer-tools/blob/master/update-version-in-file.sh)

### CODE REVIEW

File: commands/update-version-in-file/update_version_in_file.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	Args:  cobra.ExactArgs(3),
+	RunE:  run,
+}
+
+func run(cmd *cobra.Command, args []string) error {
+	toUpdateFilepath, patternFormatStr, newVersion := args[0], args[1], args[2]
+	fileToUpdateInfo, err := os.Stat(toUpdateFilepath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return stacktrace.Propagate(err, "No file exists at '%s'", toUpdateFilepath)
```
Comment (mieubrisse) [inline commands/update-version-in-file/update_version_in_file.go:35]: Good use of `''`

File: commands/update-version-in-file/update_version_in_file.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		}
+		return stacktrace.Propagate(err, "An error occurred attempting to retrieve file info for file at '%s'", toUpdateFilepath)
+	}
+	fileToUpdateMode := fileToUpdateInfo.Mode()
+	fileToUpdate, err := os.ReadFile(toUpdateFilepath)
```
Comment (mieubrisse) [inline commands/update-version-in-file/update_version_in_file.go:40]: This is a potentially big memory issue, so should have a TODO indicating the problem and the solution (switch to streaming via Scanner)

File: commands/update-version-in-file/update_version_in_file.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	if err != nil {
+		return stacktrace.Propagate(err, "An error occurred attempting to read file at '%s'", toUpdateFilepath)
+	}
+
+	if !strings.Contains(patternFormatStr, formatStrReplacementSubstr) {
```
Comment (mieubrisse) [inline commands/update-version-in-file/update_version_in_file.go:45]: Under the principle of eject-early eject-often, you'd want to do this validation earlier, before you start reading the file (which in its current incarnation is a potentially slow operation). This also lets you have a flow like:

1. Validate all your axioms first
2. Start doing logic

Rather than mixing them

File: commands/update-version-in-file/update_version_in_file.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+		return stacktrace.NewError("The replacement substring '%s' was not found in the passed '%s' as required.", formatStrReplacementSubstr, patternFormatStr)
```
Comment (mieubrisse) [inline commands/update-version-in-file/update_version_in_file.go:46]: `passed '%s'` misses semantic information (gotta assume that the person reading these log messages doesn't have the context). E.g. better would be `...was not found in the provided match regex '%s' as required`

File: commands/update-version-in-file/update_version_in_file.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	}
+
+	if !versionRegex.Match([]byte(newVersion)) {
+		return stacktrace.NewError("The version regex pattern, '%s', was not found in the provided version, '%s'", versionRegexStr, newVersion)
```
Comment (mieubrisse) [inline commands/update-version-in-file/update_version_in_file.go:50]: More accurately, the provided version doesn't match the version regex (remember that regexes are state machines so you wouldn't really say a state machine is "in" a string).

Also same thing with validation being done early, before you start doing logic

File: commands/update-version-in-file/update_version_in_file.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+)
+
+const (
+	updateVersionInFileCmdStr     = "update-version-in-file <to update filepath> <pattern format string> <new version>"
+	versionRegexStr               = "[0-9A-Za-z_./-]+"
+	formatStrReplacementSubstr    = "%s"
+	expectedNumSearchPatternLines = 1
+)
+
+var versionRegex = regexp.MustCompile(versionRegexStr)
```
Comment (mieubrisse) [inline commands/update-version-in-file/update_version_in_file.go:21]: Nice job compiling this in advance

File: commands/update-version-in-file/update_version_in_file.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+	}
+
+	if !versionRegex.Match([]byte(newVersion)) {
+		return stacktrace.NewError("The version regex pattern, '%s', was not found in the provided version, '%s'", versionRegexStr, newVersion)
+	}
+
+	searchPatternStr := fmt.Sprintf(patternFormatStr, versionRegexStr)
+	searchPatternRegex, err := regexp.Compile(searchPatternStr)
+	if err != nil {
+		return stacktrace.Propagate(err, "An error occurred creating regex pattern of %s", searchPatternStr)
```
Comment (mieubrisse) [inline commands/update-version-in-file/update_version_in_file.go:56]: Missing either `:` or `''`

File: commands/update-version-in-file/update_version_in_file.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+	}
+
+	replaceValue := fmt.Sprintf(patternFormatStr, newVersion)
+
+	numLines, err := helpers.CountLinesMatchingRegex(toUpdateFilepath, searchPatternRegex)
+	if err != nil {
+		return stacktrace.Propagate(err, "An error occurred while trying to count the number of occurrences of '%s' in '%s'", searchPatternStr, toUpdateFilepath)
+	}
+	if numLines != expectedNumSearchPatternLines {
```
Comment (mieubrisse) [inline commands/update-version-in-file/update_version_in_file.go:65]: Nice job w/named constant

File: commands/update-version-in-file/update_version_in_file.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+		return stacktrace.NewError("An incorrect amount, '%d' of lines matching '%s' was found in '%s'. '%d' matching lines were expected.", numLines, searchPatternStr, toUpdateFilepath, expectedNumSearchPatternLines)
```
Comment (mieubrisse) [inline commands/update-version-in-file/update_version_in_file.go:66]: Really good descriptive error

File: commands/update-version-in-file/update_version_in_file.go (hunk 2 above)
Comment (mieubrisse) [reply commands/update-version-in-file/update_version_in_file.go:40]: Also `fileToUpdate` isn't actually a file (which implies either a fp or filepath); it's bytes/contents of a file

File: commands/update-version-in-file/update_version_in_file.go [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+	err = os.WriteFile(toUpdateFilepath, updatedFile, fileToUpdateMode)
+	if err != nil {
+		return stacktrace.Propagate(err, "An error occurred attempting to right the updated file contents to '%s'", toUpdateFilepath)
+	}
+	return nil
+}
+
+func replaceLinesMatchingPattern(file []byte, regexPat *regexp.Regexp, replacement string) []byte {
+	lines := bytes.Split(file, []byte("\n"))
```
Comment (mieubrisse) [inline commands/update-version-in-file/update_version_in_file.go:79]: Reallyyyyy gotta be careful with the memory management on arbitrarily-huge input now; with this line we're now keeping the entire file in memory, twice. I think it's okay here as a janky/dirty thing, but just barely - should definitely have a `TODO` calling out the problem

File: commands/update-version-in-file/update_version_in_file.go [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+	for i, line := range lines {
+		if regexPat.Match(line) {
+			lines[i] = []byte(replacement)
```
Comment (mieubrisse) [inline commands/update-version-in-file/update_version_in_file.go:82]: Careful, this is now a change of functionality from the original script! Whereas the original would replace _only_ the specified regex, this now replaces the entire line.

Practically, this means that something like:
```
myVersionConst = "1.2.3"     // This gets auto-updated
```
would now get replaced to:
```
myVersionConst = "1.3.0"
```
(comment gets dropped)

What you probably want is something like `strings.Replace`

Comment (mieubrisse) [reply commands/update-version-in-file/update_version_in_file.go:82]: NOTE: This is an edge case bug, and not one that's entirely obvious. What I'd do is:
1) create a unit test to test it
2) verify that it fails
3) fix it
4) verify the unit test now passes

Now all future devs are safe against the same thing.

File: commands/update-version-in-file/update_version_in_file_test.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+import (
+	"github.com/stretchr/testify/require"
+	"regexp"
+	"testing"
+)
+
+func TestNoMatchingPatternFoundReturnsIdenticalFile(t *testing.T) {
+	replacementStr := "KURTOSIS_CORE_VERSION: string = \"0.1.3\""
+	searchPatternStr := "KURTOSIS_CORE_VERSION: string = \"[0-9A-Za-z_./-]+\""
```
Comment (mieubrisse) [inline commands/update-version-in-file/update_version_in_file_test.go:11]: Careful; you've copied in the `[0-9A-Za-z_./-]+` but that means that if we ever change it in the actual code then this will be using an oudated version. This should be dynamic

File: commands/update-version-in-file/update_version_in_file_test.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+* Something`
+
+	updatedFile := replaceLinesMatchingPattern([]byte(fileWithNoMatchingPattern), searchPatternRegex, replacementStr)
+
+	require.Equal(t, string(updatedFile), string(fileWithNoMatchingPattern))
+}
+
+func TestMatchingPatternFoundReturnsUpdatedLine(t *testing.T) {
+	replacementStr := "KURTOSIS_CORE_VERSION: string = \"0.1.3\""
+	searchPatternStr := "KURTOSIS_CORE_VERSION: string = \"[0-9A-Za-z_./-]+\""
```
Comment (mieubrisse) [inline commands/update-version-in-file/update_version_in_file_test.go:27]: Same here and everywhere

**Review by mieubrisse — APPROVED** [review]
Approved pending comments

File: commands/update-version-in-file/update_version_in_file.go [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+	"github.com/kurtosis-tech/stacktrace"
+	"github.com/spf13/cobra"
+	"os"
+	"regexp"
+	"strings"
+)
+
+const (
+	updateVersionInFileCmdStr     = "update-version-in-file <to update filepath> <pattern format string> <new version>"
+	versionRegexStr               = "[0-9A-Za-z_./-]+"
```
[the remark this answers is not in the document]
Comment (tedim52) [reply commands/update-version-in-file/update_version_in_file.go:16]: gotcha, ye i manually tested [here](https://regex101.com/) with the GoLang setting but ill add some unit tests as well
